File: Python/CodeSignal/Industry Framework/Banking System/level2.py
class BankingSystem:
    def __init__(self):
        self.accounts = {}  # account_id -> balance
        self.outgoing_transactions = {}  # account_id -> total_outgoing
    
    def create_account(self, timestamp: int, account_id: str) -> bool:
        if account_id in self.accounts:
            return False
        
        self.accounts[account_id] = 0
        self.outgoing_transactions[account_id] = 0
        return True
    
    def deposit(self, timestamp: int, account_id: str, amount: int) -> int | None:
        if account_id not in self.accounts:
            return None
        
        self.accounts[account_id] += amount
        return self.accounts[account_id]
    
    def transfer(self, timestamp: int, source_account_id: str, target_account_id: str, amount: int) -> int | None:
        # Check if both accounts exist
        if source_account_id not in self.accounts or target_account_id not in self.accounts:
            return None
        
        # Check if accounts are the same
        if source_account_id == target_account_id:
            return None
        
        # Check if source has sufficient funds
        if self.accounts[source_account_id] < amount:
            return None
        
        # Perform transfer
        self.accounts[source_account_id] -= amount
        self.accounts[target_account_id] += amount
        
        # Update outgoing transactions for source account
        self.outgoing_transactions[source_account_id] += amount
        
        return self.accounts[source_account_id]
    
    def top_spenders(self, timestamp: int, n: int) -> list[str]:
        # Get all accounts with their outgoing amounts
        spenders = []
        for account_id in self.accounts:
            outgoing = self.outgoing_transactions[account_id]
            spenders.append((account_id, outgoing))
        
        # Sort by outgoing amount descending, then by account_id ascending
        spenders.sort(key=lambda x: (-x[1], x[0]))
        
        # Take top n and format result
        result = []
        for i in range(min(n, len(spenders))):
            account_id, outgoing = spenders[i]
            result.append(f"{account_id}({outgoing})")
        
        return result

def main():
    banking = BankingSystem()
    
    # Test cases from level 2 example
    queries = [
        ("create_account", 1, "account3"),
        ("create_account", 2, "account2"),
        ("create_account", 3, "account1"),
        ("deposit", 4, "account3", 2000),
        ("deposit", 5, "account2", 3000),
        ("deposit", 6, "account3", 4000),
        ("top_spenders", 7, 3),
        ("transfer", 8, "account3", "account2", 500),
        ("transfer", 9, "account3", "account1", 1000),
        ("transfer", 10, "account1", "account2", 2500),
        ("top_spenders", 11, 3)
    ]
    
    results = []
    for query in queries:
        operation = query[0]
        if operation == "create_account":
            result = banking.create_account(query[1], query[2])
        elif operation == "deposit":
            result = banking.deposit(query[1], query[2], query[3])
        elif operation == "transfer":
            result = banking.transfer(query[1], query[2], query[3], query[4])
        elif operation == "top_spenders":
            result = banking.top_spenders(query[1], query[2])
        
        results.append(result)
        print(f"{query}: {result}")
    
    print(f"Final results: {results}")

if __name__ == "__main__":
    main()

File: Java/A-PREP/Code Signal Industry/Banking System/level2.java
import java.util.*;
import java.util.stream.Collectors;

public class BankingSystem {
    // Map to store account IDs and their balances
    private Map<String, Integer> accounts;
    // Map to track the total outgoing transactions for each account
    private Map<String, Integer> outgoingTransactions;

    // Constructor to initialize the maps
    public BankingSystem() {
        accounts = new HashMap<>();
        outgoingTransactions = new HashMap<>();
    }

    // Method to create a new account
    public boolean createAccount(int timestamp, String accountId) {
        // Check if the account already exists
        if (accounts.containsKey(accountId)) {
            return false; // Account already exists
        }
        // Initialize the account with a balance of 0 and outgoing transactions of 0
        accounts.put(accountId, 0);
        outgoingTransactions.put(accountId, 0);
        return true; // Account created successfully
    }

    // Method to deposit money into an account
    public Integer deposit(int timestamp, String accountId, int amount) {
        // Check if the account exists
        if (!accounts.containsKey(accountId)) {
            return null; // Account does not exist
        }
        // Update the account balance
        int balance = accounts.get(accountId);
        balance += amount;
        accounts.put(accountId, balance);
        return balance; // Return the new balance
    }

    // Method to transfer money between two accounts
    public Integer transfer(int timestamp, String sourceAccountId, String targetAccountId, int amount) {
        // Check if both accounts exist and are different
        if (!accounts.containsKey(sourceAccountId) || !accounts.containsKey(targetAccountId) || sourceAccountId.equals(targetAccountId)) {
            return null; // Invalid transfer
        }
        // Check if the source account has sufficient funds
        int sourceBalance = accounts.get(sourceAccountId);
        if (sourceBalance < amount) {
            return null; // Insufficient funds
        }
        // Perform the transfer
        int targetBalance = accounts.get(targetAccountId);
        sourceBalance -= amount;
        targetBalance += amount;
        accounts.put(sourceAccountId, sourceBalance);
        accounts.put(targetAccountId, targetBalance);
        // Update the outgoing transactions for the source account
        outgoingTransactions.put(sourceAccountId, outgoingTransactions.getOrDefault(sourceAccountId, 0) + amount);
        return sourceBalance; // Return the new balance of the source account
    }

    // Method to get the top N accounts with the highest outgoing transactions
    public List<String> topAccounts(int timestamp, int n) {
        // Step 1: Convert the map entries to a list
        List<Map.Entry<String, Integer>> entries = new ArrayList<>(outgoingTransactions.entrySet());

        // Step 2: Sort the list of entries
        entries.sort((a, b) -> {
            // Compare the outgoing transactions in descending order
            int cmp = b.getValue().compareTo(a.getValue());
            if (cmp == 0) {
                // If outgoing transactions are equal, compare account IDs in ascending order
                return a.getKey().compareTo(b.getKey());
            }
            return cmp; // Return the comparison result
        });

        List<String> result = new ArrayList<>();

        // Step 3: Iterate through the sorted entries and add the top N accounts to the result list
        for (int i = 0; i < Math.min(n, entries.size()); i++) {
            // Get the current entry
            Map.Entry<String, Integer> entry = entries.get(i);

            // Format the entry as "<account_id>(<outgoing_transactions>)" and add it to the result list
            result.add(entry.getKey() + "(" + entry.getValue() + ")");
        }

        // Step 5: Return the result list
        return result;
    }
}

// Main class for Level 2
public class Main {
    public static void main(String[] args) {
        // Create a new banking system
        BankingSystem bank = new BankingSystem();

        // Create accounts
        bank.createAccount(1, "account1");
        bank.createAccount(2, "account2");
        bank.createAccount(3, "account3");

        // Deposit money
        bank.deposit(4, "account1", 1000);
        bank.deposit(5, "account2", 2000);
        bank.deposit(6, "account3", 3000);

        // Transfer money
        bank.transfer(7, "account1", "account2", 500);
        bank.transfer(8, "account2", "account3", 1000);
        bank.transfer(9, "account3", "account1", 1500);

        // Get top accounts by outgoing transactions
        System.out.println(bank.topAccounts(10, 3)); // [account3(1500), account2(1000), account1(500)]
    }
}